sermant-plugins/sermant-mq-grayscale/mq-grayscale-rocketmq-plugin/src/main/java/io/sermant/mq/grayscale/rocketmq/interceptor/RocketMqSchedulerRebuildSubscriptionInterceptor.java

```diff
@@ -35,44 +39,81 @@
+            buildSql92SubscriptionData(subscriptionData, balance, topic);
+
+            // update %RETRY%+GROUP substring with sql92
+            updateRetrySubscriptionData(subscriptionData, map.values());
         }
         return context;
     }
 
-    private void buildSql92SubscriptionData(Object subscriptionData, RebalanceImpl balance) {
-        synchronized (lock) {
-            String topic = RocketMqReflectUtils.getTopic(subscriptionData);
-            if (!RocketMqSubscriptionDataUtils.getGrayTagChangeFlag(topic, balance)) {
-                return;
+    private boolean isGrayTagChanged(ConcurrentMap<String, Object> map, RebalanceImpl balance) {
+        String topic = "";
+        for (Object subscriptionData : map.values()) {
+            String tempTopic = RocketMqReflectUtils.getTopic(subscriptionData);
+            if (!tempTopic.contains(RETYPE)) {
+                topic = tempTopic;
+                break;
             }
-            String consumerGroup = balance.getConsumerGroup();
-            MQClientInstance instance = balance.getmQClientFactory();
-            if (StringUtils.isEmpty(RocketMqGrayscaleConfigUtils.getGrayGroupTag())) {
-                RocketMqConsumerGroupAutoCheck.setMqClientInstance(topic, consumerGroup, instance);
-                RocketMqConsumerGroupAutoCheck.syncUpdateCacheGrayTags();
-                RocketMqConsumerGroupAutoCheck.startSchedulerCheckGroupTask();
+        }
+        return RocketMqSubscriptionDataUtils.getGrayTagChangeFlag(topic, balance);
+    }
+
+    private void updateRetrySubscriptionData(Object subscriptionData, Collection<Object> subscriptionDatas) {
+        String originTopic = RocketMqReflectUtils.getTopic(subscriptionData);
+        for (Object subData : subscriptionDatas) {
+            String retryTopic = RocketMqReflectUtils.getTopic(subData);
+            String sqlSubstr = RocketMqReflectUtils.getSubString(subscriptionData);
+            if (retryTopic.contains(RETYPE)) {
+                RocketMqReflectUtils.getTagsSet(subData).clear();
+                RocketMqReflectUtils.getCodeSet(subData).clear();
+                RocketMqReflectUtils.setSubscriptionDatae(subData, "setExpressionType",
+                    new Class[]{String.class}, new Object[]{"SQL92"});
+                RocketMqReflectUtils.setSubscriptionDatae(subData, "setSubVersion",
+                    new Class[]{long.class}, new Object[]{System.currentTimeMillis()});
+                String originSubData = RocketMqReflectUtils.getSubString(subData);
+                RocketMqReflectUtils.setSubscriptionDatae(subData, "setSubString",
+                    new Class[]{String.class}, new Object[]{sqlSubstr});
+                LOGGER.warning(String.format(Locale.ENGLISH, "update retry topic [%s] SQL92 expression, "
+                        + "originTopic: [%s], originSubStr: [%s], newSubStr: [%s]", retryTopic, originTopic,
+                        originSubData, sqlSubstr));
             }
```

Comment:
why traverse map.values() and determine whether the topic contains again here？

Reply:
fixed